### PR TITLE
[Resolves #7366] Upgrade from Django 2.2 to 3.2

### LINKED ISSUE #7366: GNIP 88: Upgrade from Django 2.2 to 3.2
# GNIP 88: Upgrade from Django 2.2 to 3.2

## Overview

As described in this [post](https://docs.djangoproject.com/en/3.2/releases/3.2/), the support for the Django version 2.2 will end in April 2022. The new LTS version will be the 3.2

### Proposed By

@mattiagiupponi 
@giohappy 

### Assigned to Release

This proposal is for the GeoNode master.

### State

* [x] Under Discussion
* [x] In Progress
* [ ] Completed
* [ ] Rejected
* [ ] Deferred

### Motivation
Since version 2.2 will be deprecated the next year, is better to upgrade to the new TLS version.

## Proposal

Django provides a sort of [guide](https://docs.djangoproject.com/en/3.2/howto/upgrade-version/) to upgrade with some steps to follow. After this, the steps are:
- upgrade Django from 2.2 to 3.2
- upgrade GeoNode dependencies
- Fix the code that is no more compatible or deprecated (ex: `lxml ` (deprecated) or `from django.utils import six` (removed) )

## Feedback

Update this section with relevant feedbacks, if any.

## Voting

Project Steering Committee:

* Alessio Fabiani: 👍
* Francesco Bartoli:
* Giovanni Allegri: 👍
* Simone Dalmasso:
* Toni Schoenbuchner: 👍 
* Florian Hoedt: 👍

## Links

Remove unused links below.

* [Email Discussion]()
* [Pull Request]()
* [Mail Discussion]()
* [Linked Issue]()

### PR DESCRIPTION
**Pull request by mattiagiupponi**
With this PR I'm going to upgrade the current version of Django 2.2.16 with the 3.2.
This pr will fix incompatibilities between the two versions, in detail:

- Import of `django.utils.six` (Removed from Django 3) converted in `import six`
- Changing XML parsing from  `from defusedxml import lxml as dlxml` (deprecated) to `from owslib.etree import etree as dlxml` which use the standard python  `xml.etree` 
- Static files template tag changes from `{% load staticfiles %}` (deprecated) to `{% load static %}`
- Static file module from `from django.contrib.staticfiles.templatetags import staticfiles` (removed from Django 3) to `from django.templatetags.static import static` (used in Thumbnail creation)
- Fix some migration file, by substituing `on_delete='CASCASE'` with `on_delete=models.CASCADE`
- Fix `GeoNodeBaseSuiteDiscoverRunner` by adding missing parameters (DiscoverRunner updated with Django 3.2). I have added as a default value the same as the base DiscoverRunner object. If required, we can change them.
- Changing `MARKDOWNIFY` settings due to the deprecation warning
- Swap import of `OrderedDict` from `django.utils.datastructures` to standard `collections` module
- Add `DEFAULT_AUTO_FIELD="django.db.models.BigAutoField"` due to the deprecation warning provided by Django (new in 3.2) [doc](https://docs.djangoproject.com/en/3.2/ref/settings/#default-auto-field)
- `render_to_response` of `django.shortcuts` is deprecated, replaced with `render`
- `django-activity-stream=0.9.0` upgrated to `django-activity-stream=0.10.0` (Django3 support)
- `dynamic-rest` moved to master since django3 is supported `git+https://github.com/AltSchool/dynamic-rest.git@master`
- `geonode-mapstore-client` PR open here https://github.com/GeoNode/geonode-mapstore-client/pull/138 to let the library be complain with Django3

 
DEFAULT_AUTO_FIELD 
## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [ ] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [ ] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the QA checks: flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

File: geonode/maps/tests.py [hunk 1]
```diff
@@ -21,7 +21,7 @@
 import logging
```
Comment (bieganowski) [inline geonode/maps/tests.py:21]: `DEFAULT_CONTENT_TYPE` setting in this file is also depricated:
https://github.com/GeoNode/geonode/blob/c9672721ec7f974fb9465fa1ca796b6a5fa8b4e4/geonode/maps/tests.py#L662
source: https://docs.djangoproject.com/en/3.2/releases/3.0/#features-removed-in-3-0

**Review by bieganowski — CHANGES_REQUESTED** [review]
two more general comments:
- `default_app_config` variable in the `__init__.py` files of installed Geonode applications is no longer used ([deprecation note](https://docs.djangoproject.com/en/3.2/releases/3.2/#id3)), e.g.:
https://github.com/GeoNode/geonode/blob/c9672721ec7f974fb9465fa1ca796b6a5fa8b4e4/geonode/base/__init__.py#L34

- `django.core.cache.backends.memcached.MemcachedCache` backend used by default in the Geonode is deprecated, since `python-memcached` seems to be unmaintained ([deprecation note](https://docs.djangoproject.com/en/3.2/releases/3.2/#id3))
https://github.com/GeoNode/geonode/blob/c9672721ec7f974fb9465fa1ca796b6a5fa8b4e4/.env#L58

**Comment by afabiani** [comment]
@mattiagiupponi can you fix the conflicts when you have some time?

**Comment by mattiagiupponi** [comment]
> @mattiagiupponi can you fix the conflicts when you have some time?

Sure, i'm still marking some test on this branch, but i'm gonna fix the conflicts soon!

**Review by afabiani — CHANGES_REQUESTED** [review]
I have the following error:

```
(geonode) afabiani@LAPTOP-KCDVSUEA:/mnt/c/Work/geonode$ Internal Server Error: /upload/
Traceback (most recent call last):
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/contrib/auth/decorators.py", line 21, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/mnt/c/Work/geonode/geonode/decorators.py", line 284, in wrapper
    return view_or_basicauth(func, request,
  File "/mnt/c/Work/geonode/geonode/decorators.py", line 72, in view_or_basicauth
    return view(request, *args, **kwargs)
TypeError: view() missing 1 required positional argument: 'step'
"POST /upload/ HTTP/1.1" 500 169474
```

**Review by afabiani — APPROVED** [review]

**Pull request merged**